### ISSUE TITLE
Show long/lat grid lines

### ISSUE THREAD

**Issue by henriktorget** [mapbox/mapbox-gl-js#10093]
**mapbox-gl-js version**:

### Question

I am trying to add labeled latitudinal/longitudinal to my map like shown in image below. 
![image](https://user-images.githubusercontent.com/5030524/98938028-945b7880-24e7-11eb-8a36-7c6e8d5db416.png)

I can't seem to find a good way to do this from reading the documentation.

There is some implementations for this in leaflet, but the bindings for leaflet are experimental only and not suited for production.
https://github.com/mapbox/mapbox-gl-leaflet

Am I missing something? Or is this possible to make in mapbox studio or programatically with mapbox-gl?

Thanks

**Comment by mourner** [comment]
@henriktorget you can do this programmatically with a GeoJSON source, generating a bunch of lines and then using this as a source with line and label layers. Something like this:

```js
const graticule = {
    type: 'FeatureCollection',
    features: []
};
for (let lng = -170; lng <= 180; lng += 10) {
    graticule.features.push({
        type: 'Feature',
        geometry: {type: 'LineString', coordinates: [[lng, -90], [lng, 90]]},
        properties: {value: lng}
    });
}
for (let lat = -80; lat <= 80; lat += 10) {
    graticule.features.push({
        type: 'Feature',
        geometry: {type: 'LineString', coordinates: [[-180, lat], [180, lat]]},
        properties: {value: lat}
    });
}

map.on('load', () => {
    map.addSource('graticule', {
        type: 'geojson',
        data: graticule
    });
    map.addLayer({
        id: 'graticule',
        type: 'line',
        source: 'graticule'
    });
    ...
```

It might be a good idea to create a plugin for this to make it easier though.

**Comment by henriktorget** [comment]
@mourner Thanks a lot for this!

I have been trying to do something similar. 
But I recently found a dataset for different "resolution" graticule grids that i could upload to mapbox studio and toggle on zoom level. Which works well enough for now. 

If I feel that the datasets are limiting, I will adopt your idea.

**Comment by prafeb** [comment]
@henriktorget 
I am also trying to solve this. Could you please let me know what dataset you are using. I would also like to try that solution of uploading that dataset to mapbox.

Thanks for sharing this.

**Comment by letelliercle** [comment]
@henriktorget I'm also interested by the dataset solution you mentioned but I did not find any graticule grids on the web. Could you please provide me a link ? 

Thanks !

**Comment by PaparazzoKid** [comment]
> @henriktorget you can do this programmatically with a GeoJSON source, generating a bunch of lines and then using this as a source with line and label layers. Something like this:
> 
> ```js
> const graticule = {
>     type: 'FeatureCollection',
>     features: []
> };
> for (let lng = -170; lng <= 180; lng += 10) {
>     graticule.features.push({
>         type: 'Feature',
>         geometry: {type: 'LineString', coordinates: [[lng, -90], [lng, 90]]},
>         properties: {value: lng}
>     });
> }
> for (let lat = -80; lat <= 80; lat += 10) {
>     graticule.features.push({
>         type: 'Feature',
>         geometry: {type: 'LineString', coordinates: [[-180, lat], [180, lat]]},
>         properties: {value: lat}
>     });
> }
> 
> map.on('load', () => {
>     map.addSource('graticule', {
>         type: 'geojson',
>         data: graticule
>     });
>     map.addLayer({
>         id: 'graticule',
>         type: 'line',
>         source: 'graticule'
>     });
>     ...
> ```
> 
> It might be a good idea to create a plugin for this to make it easier though.

Thanks for this!! How could this be modified to show minutes as well as degrees?

**Comment by PaparazzoKid** [comment]
> ```js
> const graticule = {
>     type: 'FeatureCollection',
>     features: []
> };
> for (let lng = -170; lng <= 180; lng += 10) {
>     graticule.features.push({
>         type: 'Feature',
>         geometry: {type: 'LineString', coordinates: [[lng, -90], [lng, 90]]},
>         properties: {value: lng}
>     });
> }
> for (let lat = -80; lat <= 80; lat += 10) {
>     graticule.features.push({
>         type: 'Feature',
>         geometry: {type: 'LineString', coordinates: [[-180, lat], [180, lat]]},
>         properties: {value: lat}
>     });
> }
> 
> map.on('load', () => {
>     map.addSource('graticule', {
>         type: 'geojson',
>         data: graticule
>     });
>     map.addLayer({
>         id: 'graticule',
>         type: 'line',
>         source: 'graticule'
>     });
>     ...
> ```


Apologies for re-opening but how could this be modidied to show minutes as well as degrees?

**Comment by henriktorget** [comment]
I guess you could try something like this, where sixty minutes is one degree:

```js

one_minute = 1 / 60

for (let lng = -170; lng <= 180; lng += one_minute) {
    graticule.features.push({
        type: 'Feature',
        geometry: {type: 'LineString', coordinates: [[lng, -90], [lng, 90]]},
        properties: {value: lng}
    });
}
for (let lat = -80; lat <= 80; lat +=one_minute) {
    graticule.features.push({
        type: 'Feature',
        geometry: {type: 'LineString', coordinates: [[-180, lat], [180, lat]]},
        properties: {value: lat}
    });
}`

**Comment by PaparazzoKid** [comment]
Splendid, thanks!!

**Comment by kevando** [comment]
This is so awesome. Thanks for sharing this code. I want to use it with this example, but each time I toggle to a new style, the graticules are no longer visible. Any ideas?

https://docs.mapbox.com/mapbox-gl-js/example/setstyle/

**Comment by paulsUsername** [comment]
TLDR; you need to re add everything when you set a new style.

@kevando When you set a style you are setting a new style sheet (style.json) so the items you added to the existing stylesheet are gone (as they are set to the previous stylesheet).

**Comment by wasdee** [comment]
FYI https://observablehq.com/@listenzcc/map-grid-by-mapbox
